did-doc-diff/test/index.ts

```diff
@@ -0,0 +1,303 @@
+    '2': { a: 1, b: 2, c: 3 },
+    '3': {
+      prev: '2',
+      patches: [['put', ['d'], 4]],
+      key: 'did:key:zDnaeYgdnK7nVH2xNQENrBRJbdCQ1KKjxt5sbykiUGLAh46Ez',
+      sig: 'z3X32k2MrDtWLr56qzGrpgCdLUouZGXZLCca1RH8unoF97geJpedbWAD8pnbw2KRM7qXLupPcurTNRAdPLAVqxP7u',
+    },
+  }
+  t.deepEqual(await diffs_to_did_doc(diffs_doc), {
+    id: 'did:aic:zwulnkrxrjkobowd',
+    a: 1,
+    b: 2,
+    c: 3,
+    d: 4,
+  })
+})
+
+test('diffs_to_doc del test', async (t) => {
+  const diffs_doc = {
+    '2': { a: 1, b: 2, c: 3 },
+    '3': {
+      prev: '2',
+      patches: [['del', ['b']]],
+      key: 'did:key:zDnaeYgdnK7nVH2xNQENrBRJbdCQ1KKjxt5sbykiUGLAh46Ez',
+      sig: 'zsXdHvXFyrZQ5dtmtaEcPVbwSgMFZWznp3V8SJGUEVPWengoo1N4tsDgcvC6e2ZwiHvWekdjLmo2o2XLJfoErstN',
+    },
+  }
+  t.deepEqual(await diffs_to_did_doc(diffs_doc), {
+    id: 'did:aic:zwulnkrxrjkobowd',
+    a: 1,
+    c: 3,
+  })
+})
+
+test('diffs_to_did_doc skip bad prev test', async (t) => {
+  const diffs_doc = {
+    '2': { a: 1, b: 2, c: 3 },
+    '3': {
+      prev: '1',
+      patches: [['del', ['b']]],
+      key: 'did:key:zDnaeYgdnK7nVH2xNQENrBRJbdCQ1KKjxt5sbykiUGLAh46Ez',
+      sig: 'z3wZ4j3HoKaAPfv1eTTeknCYsUbjfkbzs7qQkyaqVqBi4iNbvA16A9totDJfRaZZ5PxVwHrFBgCs4BGvns6a5TAp',
+    },
+  }
+  t.deepEqual(await diffs_to_did_doc(diffs_doc), {
+    id: 'did:aic:zwulnkrxrjkobowd',
+    a: 1,
+    b: 2,
+    c: 3,
+  })
+})
+
+test('diffs_to_did_doc test key gen', async (t) => {
+  //gen key
+  const key = await crypto.EcdsaKeypair.create({ exportable: true })
+  // console.log(await key.export('base58btc'), key.did())
+
+  // make message
+  const data_object = {
+    prev: 'tid',
+    patches: [
+      ['put', ['path'], 'value'],
+      ['del', ['path']],
+    ],
+    key: key.did(),
+    sig: '',
+  }
+  const signed = await sign(data_object, key) // replace sig with valid signature
+  const msg = JSON.stringify(signed)
+
+  // parse message
+  const recv = JSON.parse(msg)
+  if (recv.sig[0] !== 'z') {
+    throw "signatures must be 'z'+ base58btc"
+  }
+  const recv_sig = uint8arrays.fromString(recv.sig.slice(1), 'base58btc')
+  recv.sig = ''
+  const recv_data = uint8arrays.fromString(JSON.stringify(recv))
+
+  // validate signature
+  const valid = await crypto.verifyEcdsaSig(recv_data, recv_sig, recv.key)
+  t.assert(valid)
+})
+
+test('sign validate_sig', async (t) => {
+  // pre baked signed diff
+  t.assert(
+    await validate_sig({
+      prev: '3j2b-pul-dlia-22',
+      patches: [],
+      key: 'did:key:zDnaeYgdnK7nVH2xNQENrBRJbdCQ1KKjxt5sbykiUGLAh46Ez',
+      sig: 'z5wYnrLa136SZsdxEvrbpLwu5wi6FDsynwiVeNR3KoJqVzgBUXLv9VQffLaz78w4tngKURh1cdbZ837S6PFC9Tu4y',
+    }),
+  )
+
+  // fresh baked signed diff
+  const diff = await sign(
+    {
+      prev: '3j2b-pul-dlia-22',
+      patches: [],
+      key: key_account.did(),
+      sig: '',
+    },
+    key_account,
+  )
+  // console.log(diff)
+  t.assert(await validate_sig(diff))
+})
+
+test('tid next', async (t) => {
```

Comment:
this is a good test to have, but maybe we put it in the tid test file in `common`?

Reply:
Yes I planed to move it but was useing to test that I could import from module to module I can move it now.